### ISSUE TITLE
Can't cancel removal of a block or container in layout by setting remove attribute value to false

### ISSUE THREAD

**Issue by thomasnordkvist** [magento/magento2#1931]
I inherit the blank theme and i wanted to have the top.links visible in the checkout so i created this file:

Vendor/Theme/Magento_Checkout/layout/checkout_index_index.xml
containing this:

```
<?xml version="1.0"?>
<!--
/**
 * Copyright © 2015 Magento. All rights reserved.
 * See COPYING.txt for license details.
 */
-->
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" layout="2columns-left" xsi:noNamespaceSchemaLocation="../../../../../../../lib/internal/Magento/Framework/View/Layout/etc/page_configuration.xsd">
    <body>
        <referenceContainer name="header.panel" remove="false"/>
    </body>
</page>
```

Reading [This](http://devdocs.magento.com/guides/v2.0/frontend-dev-guide/layouts/xml-instructions.html#fedg_layout_xml-instruc_ex_ref) i understand, that it sould now be possible to cancel a removal instruction with the above instruction, but it doesn't seem to work.

I know the file is parsed, and included in the layout, because if i add a block from this file it will show up on the checkout page.

/Thomas

**Comment by daim2k5** [comment]
The GitHub issue tracker is intended for technical issues only. Please refer to the [Community Forums](https://community.magento.com) or [Magento Stack Exchange](http://magento.stackexchange.com/) site for technical questions. In your case, the [programming questions forum](https://community.magento.com/t5/Programming-Questions/bd-p/programming-magento-2) is likely the most appropriate. Feel free to reopen this issue if you think you have encountered a bug in Magento 2.

**Comment by philippsander** [comment]
this sounds like a technical issue for me... I have the same problem.

EDIT: the documentation clearly says the the removeal process is canceled. that does not work!

**Comment by thomasnordkvist** [comment]
I had almost forgotten this issue.
It should be reopened as a bug.

I did some copy paste to build up the parts i wanted back in the header.

**Comment by AntonEvers** [comment]
This problem can be reproduced by removing an element in a parent theme and trying to cancel it in a child theme.

**Comment by BenGlassAtlas** [comment]
Any news on this yet Magento? It doesn't even look it's been been internally raised or acknowledged yet.

CC @daim2k5

**Comment by dexterchua19** [comment]
Any news about this? I'm developing a website and this same problem occurs in me.

**Comment by philippsander** [comment]
anything new?

**Comment by pboisvert** [comment]
@tkacheva can you ask the team to investigate this one?

**Comment by tkacheva** [comment]
@eug123 please take a look MAGETWO-54935

**Comment by philippsander** [comment]
come on guys. almost 4 months to fix this?

**Comment by tkacheva** [comment]
@philippsander issue is waiting to be merged

**Comment by Styopchik** [comment]
Still waiting

**Comment by torreytsui** [comment]
### A limited workaround for the time being

If the purpose is simply hiding and displaying specific blocks and if you have full control on those layouts, you may interest in this limited workaround.

```diff
<!-- parent_layout_handle.xml -->
+ <referenceContainer name="root" display="false"/>
- <referenceContainer name="root" remove="true"/>

<!-- child_layout_handle.xml -->
+ <referenceContainer name="root" display="true"/>
- <referenceContainer name="root" remove="false"/>
```

Use `display` instead of `remove`. This workaround is very limited as we don't usually have full controls on all layout. For example, many Magento vanilla handles have `remove="true"` instructions and we basically cannot add those blocks back before a fix is released.

### `remove="false"` seems not implemented

If you do need the `remove="false"` to work before a fix is released, there is a way of hacking.

```php
// \Magento\Framework\View\Layout\Reader\Block::scheduleReference
protected function scheduleReference(
        Layout\ScheduledStructure $scheduledStructure,
        Layout\Element $currentElement
    ) {
        $elementName = $currentElement->getAttribute('name');
        $elementRemove = filter_var($currentElement->getAttribute('remove'), FILTER_VALIDATE_BOOLEAN);
        if ($elementRemove) {
            $scheduledStructure->setElementToRemoveList($elementName);
        } else {
            $data = $scheduledStructure->getStructureElementData($elementName, []);
            $data['attributes'] = $this->mergeBlockAttributes($data, $currentElement);
            $this->updateScheduledData($currentElement, $data);
            $this->evaluateArguments($currentElement, $data);
            $scheduledStructure->setStructureElementData($elementName, $data);
        }
    }
```

I suppose `$scheduledStructre->unsetElementFromListToRemove($elementName)` should be called in the `else` condition or something relevant so that the remove flag of the block will be reset and the block can be kept.

```diff
        if ($elementRemove) {
            $scheduledStructure->setElementToRemoveList($elementName);
        } else {
            $data = $scheduledStructure->getStructureElementData($elementName, []);
            $data['attributes'] = $this->mergeBlockAttributes($data, $currentElement);
            $this->updateScheduledData($currentElement, $data);
            $this->evaluateArguments($currentElement, $data);
            $scheduledStructure->setStructureElementData($elementName, $data);
+            $scheduledStructre->unsetElementFromListToRemove($elementName)
        }
```

Since there seems to be a fix waiting to merge, I'm not going to raise a PR.

**Comment by KrystynaKabannyk** [comment]
hi @thomasnordkvist, the issue has been fixed under MAGETWO-54935 in develop branch, you can find it here and apply for your version: https://github.com/magento/magento2/commit/9bd170df9fb459cd3d5fb6fcc76cd1e3ce63919f

**Comment by PieterCappelle** [comment]
Please merge this in some upcoming release. Thanks.

**Comment by korostii** [comment]
Hi @KrystynaKabannyk, could you please reopen this issue (until the fix appears in a released 2.1.x version)?
Thanks in advance.

**Comment by KrystynaKabannyk** [comment]
hi @korostii, this issue hasn't been scheduled to be backported, it will go to 2.2.0 release.

**Comment by korostii** [comment]
@KrystynaKabannyk
I hate to underline the obvious, but if the bug was discovered on 2.1.x and then reportedly "fixed", it is commonly expected to see that fix in the closest patch (2.1.x) release. 

Considering that for many of the merchants the switch to 2.2 will probably be somewhat painful, resource-heavy (see this marvelous [list of backwards-incompatible changes](https://github.com/magento/devdocs/blob/develop/guides/v2.2/release-notes/backward-incompatible-changes.md) for a taste) and delayed for at least a couple more months, it would make sense to have it on the 2.1.x branch as well. Furthermore, someone from the community just might want this fix badly enough and make pull request aiming `2.1-develop`.

Either that, or some disclaimer saying that "this is a known issue that won't ever be fixed in 2.1. Please move along to 2.2". You do know that there are people re-testing every bug they have encountered earlier after each new release in order to see if it got fixed? right?

Anyhow, having the issue marked as "closed" (while there isn't any straightforward way to use it available to ordinary users) is very misleading. Could you at least leave the issue open until 2.2.0 gets released?

**Comment by KrystynaKabannyk** [comment]
hi @korostii, we reviewed the issue again and made the decision to backport it into 2.1.x.  Internal ticket for it is created MAGETWO-69715. Thank you for raising this issue again!

**Comment by PieterCappelle** [comment]
Thank god for people like @korostii. Thanks for the effort @KrystynaKabannyk.

**Comment by magento-engcom-team** [comment]
The issue already fixed and will be available in release 2.2.0 very soon

**Comment by korostii** [comment]
Hi @magento-engcom-team, 
I am having a bad case of deja vu and I just hope that this message above was an automated one and someone has just missed the discussion above (as opposed to outright ignoring it and turning this thing over again).

The issue is reported and reproduced on 2.1.x
The issue isn't fixed on 2.1.x
How come it's "closed" already (again) ?!

Just trying to make sense of it, please help me understand the issue flow here.
I mean, this isn't your internal JIRA where the only thing that matters is whether is fix is implemented already. People here care whether they can update to the _latest release_ in order to have the bug gone (and they can't, in this case)

**Comment by Morgon** [comment]
@magento-engcom-team @KrystynaKabannyk - Magento committed to backporting this into 2.1 and has marked the ticket as Done (even after @korostii mentioned that Magento may be misunderstanding). However, it's not even in 2.1.10-dev. Please fix the status of this report and update the community on an ETA.

**Comment by quisse** [comment]
Still not backported to 2.1.*?

**Comment by magento-engcom-team** [comment]
Hi @thomasnordkvist. Thank you for your report. 
The issue has been fixed in magento/magento2#14198 by @quisse in 2.1-develop branch
Related commit(s):
 - [44746e008a9563869af7f02d0e02dd36dbd0f82e](https://github.com/magento/magento2/commit/44746e008a9563869af7f02d0e02dd36dbd0f82e)
 - [2423facb0683e4b2a1cb64aadb80629e8c7865b0](https://github.com/magento/magento2/commit/2423facb0683e4b2a1cb64aadb80629e8c7865b0)

The fix will be available with the upcoming 2.1.14 release.

**Comment by quisse** [comment]
@magento-engcom-team Do you mean 2.1.13 or is that one going to be skipped?
